### ISSUE TITLE
Bug in AOT mode with Angular 5 when using yarn:  TypeScript compilation failed

### ISSUE THREAD

**Issue by quanterion** [angular/angular-cli#8107]
### Versions.
Yarn: 1.2.1
Angular CLI: 1.5.0-rc.1
Node: 8.4.0
OS: win32 x64
Angular: 5.0.0-rc.3
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

angular/cli: 1.5.0-rc.1
angular-devkit/build-optimizer: 0.0.28
angular-devkit/core: 0.0.20
angular-devkit/schematics: 0.0.34
ngtools/json-schema: 1.1.0
schematics/angular: 0.0.46
typescript: 2.4.2
webpack: 3.7.1

### Repro steps.

1. ng set --global packageManager=yarn
2. ng new myApp
3. update package.json to Angular 5
4. ng serve --aot

### The log given by the failure.
ERROR in ./src/main.ts
Module build failed: Error: TypeScript compilation failed.
    at plugin.done.then (X:\ng15y\node_modules\@angular\cli\node_modules\@ngtools\webpack\src\loader.js:479:27)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
ERROR in ./src/polyfills.ts
Module build failed: Error: TypeScript compilation failed.
    at plugin.done.then (X:\ng15y\node_modules\@angular\cli\node_modules\@ngtools\webpack\src\loader.js:479:27)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 @ multi ./src/polyfills.ts
ERROR in

### Mention any other details that might be useful.
I use yarn 1.2.1 to install dependencies. I couldn't reproduce this error in empty project with npm 5, but after installing libraries material-beta.12, covalent-beta.8-1 and angulartics-2.4.0 the same problem also appears.
Reverting project to Angular 4.4.5 fixes this issue in both cases.

**Comment by quanterion** [comment]
Project example: https://github.com/quanterion/cli-aot-bug

**Comment by quanterion** [comment]
I have Angular-Cli installed via npm when I remove cli from npm and install via yarn it gives mo detailed error message when serving with ng serve --aot:

ERROR in ./src/main.ts
Module build failed: Error: TypeScript compilation failed.
    at plugin.done.then (X:\ng15\node_modules\@angular\cli\node_modules\@ngtools\webpack\src\loader.js:479:27)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
ERROR in ./src/polyfills.ts
Module build failed: Error: TypeScript compilation failed.
    at plugin.done.then (X:\ng15\node_modules\@angular\cli\node_modules\@ngtools\webpack\src\loader.js:479:27)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 @ multi ./src/polyfills.ts
ERROR in

webpack: Failed to compile.
X:\ng15\node_modules\typescript\lib\typescript.js:69795
                throw e;
                ^

TypeError: Cannot read property 'length' of undefined
    at Object.unescapeIdentifier (X:\ng15\node_modules\typescript\lib\typescript.js:10550:26)
    at getTextOfNode (X:\ng15\node_modules\typescript\lib\typescript.js:68575:27)
    at emitIdentifier (X:\ng15\node_modules\typescript\lib\typescript.js:66931:19)
    at pipelineEmitUnspecified (X:\ng15\node_modules\typescript\lib\typescript.js:66594:28)
    at pipelineEmitWithHint (X:\ng15\node_modules\typescript\lib\typescript.js:66566:50)
    at pipelineEmitWithSourceMap (X:\ng15\node_modules\typescript\lib\typescript.js:66558:17)
    at emitNodeWithComments (X:\ng15\node_modules\typescript\lib\typescript.js:64200:17)
    at pipelineEmitWithComments (X:\ng15\node_modules\typescript\lib\typescript.js:66547:17)
    at pipelineEmitWithNotification (X:\ng15\node_modules\typescript\lib\typescript.js:66541:17)
    at emit (X:\ng15\node_modules\typescript\lib\typescript.js:66528:13)

**Comment by pschmoek** [comment]
Same here.
It's because yarn fails to dedupe typescript:

package.json
```
{
  "dependencies": {
    "@angular/core": "5.0.0-rc.3",
    ...
  },
  "devDependencies": {
    "@angular/cli": "1.5.0-rc.1",
    "typescript": "~2.4.2",
    ...
  }
}
```

yarn.lock
```
"typescript@>=2.0.0 <2.6.0", typescript@^2.3.3:
  version "2.5.3"
  resolved "URL"

typescript@~2.4.2:
  version "2.4.2"
  resolved "URL"
```

Using npm only 2.4.2 is used and build --prod --aot works fine.

**Comment by pschmoek** [comment]
... tested with npm 3.10.10, dunno about 5. Tired of yarn bugs :sleepy:

**Comment by Matmo10** [comment]
I'm experiencing the same error. This was happening on the 1.5.0 beta releases also, I just never filed an issue because I figured something was wrong on my end.  

yarn.lock has been my source of truth for awhile now, switching to npm scares me 😨

**Comment by csentis** [comment]
@Matmo10 same here, although [npm‘s package lockfile](https://docs.npmjs.com/files/package-locks) seems tempting...

**Comment by devoto13** [comment]
It is not a Yarn bug anyhow. Hoisting is an implementation detail of the package manager (including Yarn) and if one wants to have single instance of some package, they should specify it as a peer dependency instead of a normal dependency. Hoping that scope hoisting will hoist it to the single package is wrong.

So the right fix here is probably to change `@angular/devkit` package to have `typescript` as a peer dependency instead of a normal dependency. And this will be correct requirement for a package manager and will ensure that dependency is present only once in the `node_modules` (which looks like the root cause of the exception).

cc @filipesilva

PS There were several fixes in Yarn around peer dependencies resolution (https://github.com/yarnpkg/yarn/issues/4433 and https://github.com/yarnpkg/yarn/issues/4539) in recent versions, so I would recommend to use >1.2.1 which contains all of them .

**Comment by filipesilva** [comment]
We're aware of this problem but there actually is no non-breaking change that would fix it.

It's not about the peer dependency on build-optimizer. Build optimizer doesn't need to use the same typescript as your project.

It's about peer dependency that `@ngtools/webpack` has on typescript, and the fact that we use the same npm package for both global and local installs.

`@ngtools/webpack` has a peer dependency on typescript to ensure it uses the project TS. But since we use the same global and local package, this is a problem. On your project, `@ngtools/webpack` wants to use the local typescript. But on the global install there is no project typescript. 

So to prevent peer dependency errors/warnings `@angular/cli` provides TS as well in its dependencies. And we rely on package manager hoisting to ensure the right dependency is provided.

This is wrong because, as @devoto13 said, to hoist or not is an implementation detail of the package manager. The right way to do it is with peer dependencies.

But we can't do that until we split the global and local CLI installs. And that's a breaking change. So that's where we stand right now. There may be a couple of hacks we do to try to sort this out for now but they are hacks.

**Comment by filipesilva** [comment]
Fixed by https://github.com/angular/angular-cli/pull/8185

**Comment by Splaktar** [comment]
@filipesilva Thank you! I can indeed build now with the latest master, but I'm seeing the following warnings, are these expected or should another issue be raised?

```
WARNING in ./node_modules/@angular/core/esm5/core.js
6436:15-36 Critical dependency: the request of a dependency is an expression
    at ImportLazyContextDependency.getWarnings (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/dependencies/ContextDependency.js:39:18)
    at Compilation.reportDependencyErrorsAndWarnings (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:703:24)
    at Compilation.finish (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:561:9)
    at applyPluginsParallel.err (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compiler.js:506:17)
    at /Users/splaktar/Git/tf/webapp/node_modules/tapable/lib/Tapable.js:289:11
    at _addModuleChain (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:507:11)
    at processModuleDependencies.err (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:477:14)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
 @ ./node_modules/@angular/core/esm5/core.js
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts

WARNING in ./node_modules/@angular/core/esm5/core.js
6456:15-102 Critical dependency: the request of a dependency is an expression
    at ImportLazyContextDependency.getWarnings (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/dependencies/ContextDependency.js:39:18)
    at Compilation.reportDependencyErrorsAndWarnings (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:703:24)
    at Compilation.finish (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:561:9)
    at applyPluginsParallel.err (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compiler.js:506:17)
    at /Users/splaktar/Git/tf/webapp/node_modules/tapable/lib/Tapable.js:289:11
    at _addModuleChain (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:507:11)
    at processModuleDependencies.err (/Users/splaktar/Git/tf/webapp/node_modules/webpack/lib/Compilation.js:477:14)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
 @ ./node_modules/@angular/core/esm5/core.js
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts
```

**Comment by Splaktar** [comment]
While I can build, I can't run my Angular v5 app with the CLI code in master due to this:
```
TypeError: querystring.unescape is not a function
    at getFilenameFromUrl (/Users/splaktar/Git/tf/webapp/node_modules/webpack-dev-middleware/lib/GetFilenameFromUrl.js:33:21)
    at module.exports (/Users/splaktar/Git/tf/webapp/node_modules/webpack-dev-middleware/lib/GetFilenameFromUrl.js:62:9)
    at webpackDevMiddleware (/Users/splaktar/Git/tf/webapp/node_modules/webpack-dev-middleware/middleware.js:40:18)
    at Layer.handle [as handle_request] (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:317:13)
    at /Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:275:10)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:127:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at Server.setContentHeaders (/Users/splaktar/Git/tf/webapp/node_modules/webpack-dev-server/lib/Server.js:496:3)
    at Layer.handle [as handle_request] (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:137:13)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at Route.dispatch (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/layer.js:95:5)
    at /Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:281:22
    at param (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:354:14)
    at param (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:365:14)
    at Function.process_params (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:410:3)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:275:10)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:127:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at Server.app.all (/Users/splaktar/Git/tf/webapp/node_modules/webpack-dev-server/lib/Server.js:77:47)
    at Layer.handle [as handle_request] (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:137:13)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:131:14)
    at Route.dispatch (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/layer.js:95:5)
    at /Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:281:22
    at param (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:354:14)
    at param (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:365:14)
    at Function.process_params (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:410:3)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:275:10)
    at expressInit (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/middleware/init.js:40:5)
    at Layer.handle [as handle_request] (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:317:13)
    at /Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:275:10)
    at query (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/middleware/query.js:45:5)
    at Layer.handle [as handle_request] (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:317:13)
    at /Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:275:10)
    at Function.handle (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/router/index.js:174:3)
    at EventEmitter.handle (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/application.js:174:10)
    at Server.app (/Users/splaktar/Git/tf/webapp/node_modules/express/lib/express.js:39:9)
    at Server.new_handler (/Users/splaktar/Git/tf/webapp/node_modules/sockjs/lib/utils.js:89:20)
    at emitTwo (events.js:106:13)
    at Server.emit (events.js:191:7)
    at HTTPParser.parserOnIncoming [as onIncoming] (_http_server.js:546:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23)
```

**Comment by blingerson** [comment]
@Splaktar, were you able to fix that warning?

**Comment by Matmo10** [comment]
@blingerson This is already fixed in the last release or 2 of 1.5.x

**Comment by Splaktar** [comment]
@blingerson yes, it's fixed in one of the RCs after 3. Working in the `1.5.0` release.

**Comment by ThalesMatoso** [comment]
I still have this error here, how do I correct it?

'' Critical dependency: the request of a dependency is an expression
     at ImportLazyContextDependency.getWarnings ''

**Comment by SFDanny** [comment]
@ThalesMatoso 
you might need this line into webpack.config.js

```
new webpack.ContextReplacementPlugin(
            // The (\\|\/) piece accounts for path separators in *nix and Windows
            /angular(\\|\/)core(\\|\/)esm5/,
            path.resolve(__dirname, "app"), // location of your src
            {} // a map of your routes
        ),
```

**Comment by ThalesMatoso** [comment]
@SFDanny thank you, but this is not resolve my problem, after search on much on web,  i discovered that it was error on library ngx-mask.

following on issue for more details.
[https://github.com/JsDaddy/ngx-mask/issues/49](url)
